src/v/storage/index_state.h

```diff
@@ -12,28 +12,156 @@
+          std::end(_relative_offset_index),
+          std::forward<Fn>(fn));
+    }
+
+    template<class Fn>
+    void for_each_relative_time_index(Fn&& fn) const {
+        assert_column_sizes();
+        std::for_each(
+          std::begin(_relative_time_index),
+          std::end(_relative_time_index),
+          std::forward<Fn>(fn));
+    }
+
+    template<class Fn>
+    void for_each_position_index(Fn&& fn) const {
+        assert_column_sizes();
+        std::for_each(
+          std::begin(_position_index),
+          std::end(_position_index),
+          std::forward<Fn>(fn));
+    }
+
+    void assert_column_sizes() const {
+        auto ro = _relative_offset_index.size();
+        auto rt = _relative_time_index.size();
+        auto ps = _position_index.size();
+        vassert(
+          ro == rt && rt == ps, "Column sizes differ: {}, {}, {}", ro, rt, ps);
+    }
+
+    /// Return index of the element or nullopt
+    std::optional<int> offset_lower_bound(uint32_t needle) const noexcept;
+
+    /// Return index of the element or nullopt
+    std::optional<int> position_upper_bound(uint64_t needle) const noexcept;
+
+    /// Return index of the element or nullopt
+    std::optional<int> time_lower_bound(uint32_t needle) const noexcept;
+
+    /// If the size() ==  1 reset the time column with the
+    /// provided value.
+    /// If the relative_time_index column is empty or size() > 1
+    /// the operation fails and method returns 'false'.
+    bool try_reset_relative_time_index(uint32_t);
+
+    bool empty() const noexcept;
+    size_t size() const noexcept;
+
+    // These methods are used by serialization
+    chunked_vector<uint32_t> copy_relative_offset_index() const noexcept;
+    chunked_vector<uint32_t> copy_relative_time_index() const noexcept;
+    chunked_vector<uint64_t> copy_position_index() const noexcept;
+    void assign_relative_offset_index(chunked_vector<uint32_t>) noexcept;
+    void assign_relative_time_index(chunked_vector<uint32_t>) noexcept;
+    void assign_position_index(chunked_vector<uint64_t>) noexcept;
+
+    void
+    add_entry(uint32_t relative_offset, uint32_t relative_time, uint64_t pos);
+
+    /// Pop back one element. This is ineffective with columnar format but
+    /// it's not invoked often and when it is invoked it usually invoked not
+    /// that many times.
+    void pop_back(int n = 1);
+
+    void shrink_to_fit();
+
+    /// Make deep copy
+    compressed_index_columns copy() const;
+
+    friend bool operator==(
+      const compressed_index_columns& lhs, const compressed_index_columns& rhs);
+
+    friend std::ostream&
+    operator<<(std::ostream&, const compressed_index_columns&);
+
+private:
+    using pos_column_t = deltafor_column<
+      uint64_t,
+      details::delta_delta<uint64_t>,
+      position_index_step>;
+    using column_t = deltafor_column<uint64_t, details::delta_xor, 0>;
+
+    column_t _relative_offset_index;
+    column_t _relative_time_index;
```

Comment:
during pr review @Lazin mentioned that the types here might be incorrect as timestamps are not always monotonic

not sure what he meant but leaving this comment so he doesn't forget to re-review this part

Reply:
good point. does the index make that assumption? i would have assumed that the delta was signed.

Reply:
I decided to keep them as is for now. We can always switch to another algorithm. This only affects the way we store data in-memory. The upside is that xor-based delta encoding can't be broken by wrong data.